gnrc/lorawan: add basic LoRaWAN 1.1 features

Ollrogge opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md.
-->

### Contribution description
This PR adds basic support for LoRaWAN version 1.1. This includes joining a network and sending / receiving data.
This PR also modifies the LoRaWAN 1.0 code to use the 1.1 terms. This should not change the behavior of the 1.0 code and everything should still work.

Not implemented are ReJoin-Requests and MAC commands such as RekeyInd and RekeyConf.

For testing purposes I setup a private LoRaWAN network using the [Chirpstack](https://www.chirpstack.io/) and was able to join the network and send data.

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure
* `tests/gnrc_lorawan_11`
* `examples/gnrc_lorawan` if able to setup a private LoRaWAN network

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

jia200x commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan.c:
```diff
@@ -76,14 +76,34 @@ static void _sleep_radio(gnrc_lorawan_t *mac)
 
-void gnrc_lorawan_init(gnrc_lorawan_t *mac, uint8_t *nwkskey, uint8_t *appskey)
+#if IS_ACTIVE(CONFIG_LORAWAN_1_1)
+void gnrc_lorawan_init_11x(gnrc_lorawan_t *mac, uint8_t *joineui,
```
I think at this point it makes sense to use a common `gnrc_lorawan_init` function but with a "ctx" parameter. E.g
```c
void gnrc_lorawan_init(gnrc_lorawan_t *mac, const gnrc_lorawan_ctx_t *ctx)
{
   mac->nwkskey = ctx->nwskey;
    ....
}
```
Since this function would be resolved at compile time (e.g user cannot switch between 1.0.3 and 1.1 on runtime), I think it will simplify things a bit

Ollrogge replied:
Is there a kind of hard cut between `gnrc_netif_lorawan` and `gnrc_lorawan` ? Otherwise I could just pass in a pointer to `gnrc_netif_lorawan_t` without needed to declare a new ctx struct ?

Ollrogge replied:
Okay I see :) I added the `gnrc_lorawan_ctx_t` struct.

Another question regarding `gnrc netif`: Right now to configure the NwkSKey for Lorawan 1.0x one has to set `CONFIG_LORAMAC_FNWKSINT_KEY_DEFAULT`. As I think this is kind of confusing for someone wanting to configure a 1.0x node I think it would be better to use the old 1.0x terms in the netif layer and only switch to the new terms internally ?

jia200x replied:
hmmm good question... I usually tend to think that's better to reflect the actual key names depending on the version. For example, a user that explicitly wants to use LoRaWAN 1.1 will probably know the key names. Most LoRaWAN users will still use LoRaWAN 1.0.x, so it's probably to keep the original names there. You can always fallback to macro aliases that use the terms internally.

jia200x commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_crypto.c:
```diff
@@ -55,25 +97,104 @@ void gnrc_lorawan_calculate_join_mic(const uint8_t *buf, size_t len,
-void gnrc_lorawan_calculate_mic(const le_uint32_t *dev_addr, uint32_t fcnt,
-                                uint8_t dir, iolist_t *frame,
-                                const uint8_t *nwkskey, le_uint32_t *out)
+#if IS_USED(CONFIG_LORAWAN_1_1)
```
Same here. You can merge these two functions and use a common signature. If you have different parameters, consider a structure.
You can also kind of merge the logic here: `gnrc_lorawan_calculate_mic_uplink_11x` is (almost) a subset of `gnrc_lorawan_calculate_mic_uplink_10x`.

Also, you can use a regular `if` with `IS_ACTIVE`or `IS_USED` in some parts. The compiler will remove branches that are never reached. Of course, you cannot do that with struct members that don't exist on compile time (does who appear only with LoRaWAN 1.1).

For instance, you could write this functions as:
```c
void gnrc_lorawan_calculate_mic_uplink(const le_uint32_t *dev_addr, uint32_t fcnt,
                                           ... lorawan_mic_t *ctx, ...
                                           le_uint32_t *out)
{
    orawan_block0_t block0 = { 0 };

    block0.fb = MIC_B0_START;
    block0.dir = 0x00;
        
    memcpy(&block0.dev_addr, dev_addr, sizeof(le_uint32_t));

    block0.fcnt = byteorder_htoll(fcnt);
    block0.len = iolist_size(frame);

    /* cmacF = aes128_cmac(FNwkSIntKey, B0 | msg) */
    cmac_init(&CmacContext, fnwksintkey, LORAMAC_FNWKSINTKEY_LEN);
    cmac_update(&CmacContext, &block0, sizeof(block0));
    for (iolist_t *io = frame; io != NULL; io = io->iol_next) {
        cmac_update(&CmacContext, io->iol_base, io->iol_len);
    }
    cmac_final(&CmacContext, digest);
  
    if (IS_ACTIVE(GNRC_LORAWAN_1_1)) {
        /* Do 1.1 specific stuff an get specific keys from `ctx`*/
    }
}
```

Ollrogge replied:
When exactly can I use `if` with this macro ? I tried to e.g. use [here](https://github.com/Ollrogge/RIOT/blob/74caa2caa3a125d90b67124481d5ec46c57f5e16/sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c#L277) and it didn't work.
I find it kinda weird to depend on compiler optimizations to make my code compile ...

jia200x replied:
it probably failed because `_appkey` exists only when you set LoRaWAN 1.1. Since the "topology" of the file is the same, the variables should always exist.

You can always use `if` with this macro, since always resolves to 0 or 1.

For example:
```c
/* foo */
if (IS_ACTIVE(MY_MODULE)) {
    /* bar */
}
```
In this example the snippets "foo" and "bar" are compiled like regular C code. But in case "MY_MODULE" is not used, the compiler will optimize out "bar". The compiler will still find syntax errors in "bar", since the compiler parses the snippet anyway.

jia200x replied:
> I find it kinda weird to depend on compiler optimizations to make my code compile ...

What do you mean with that? You don't depend on compiler optimizations to compile your code (if you turn off compiler optimizations it will still compile without issues). Could you elaborate more?

Ollrogge replied:
> it probably failed because `_appkey` exists only when you set LoRaWAN 1.1. Since the "topology" of the file is the same, the variables should always exist.
> 
> You can always use `if` with this macro, since always resolves to 0 or 1.
> 
> For example:
> 
> ```c
> /* foo */
> if (IS_ACTIVE(MY_MODULE)) {
>     /* bar */
> }
> ```
> 
> In this example the snippets "foo" and "bar" are compiled like regular C code. But in case "MY_MODULE" is not used, the compiler will optimize out "bar". The compiler will still find syntax errors in "bar", since the compiler parses the snippet anyway.

Having `_appkey` exist all the time fixed the issue. Thank you ! 
I misunderstood how using `if (IS_USED(...)` works. Your example and explanation cleared things up !

jia200x commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_crypto.c:
```diff
@@ -136,30 +257,89 @@ void gnrc_lorawan_decrypt_join_accept(const uint8_t *key, uint8_t *pkt,
-                                        const uint8_t *dev_nonce,
-                                        const uint8_t *appkey, uint8_t *nwkskey,
-                                        uint8_t *appskey)
+#if IS_ACTIVE(CONFIG_LORAWAN_1_1)
```
If you can join these two functions it would be also nice

Ollrogge replied:
Can you explain to me why joining functions and using a struct as parameter that has different members depending on the LoRaWAN version used is a better style than just using two functions ? 
To me having 2 separate functions is way cleaner.

Ollrogge replied:
+ Using two functions avoids having to initialize a struct that takes up stack space.

jia200x replied:
> To me having 2 separate functions is way cleaner.

To be clear, I'm not arguing for using a struct with different elements rather than two functions. Of course two functions are better when the functions diverge. But in this case, LoRaWAN 1.0.x is still a subset of LoRaWAN 1.1.
Therefore, these function cannot be _that_ divergent (and they don't seem to diverge either).
If you think there are some snippets that don't make sense at all to merge, it is fine. But I think we should try to avoid having the same snippets spread on different functions if they perform almost the same task.

jia200x commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mcps.c:
```diff
@@ -264,18 +280,41 @@ size_t gnrc_lorawan_build_uplink(gnrc_lorawan_t *mac, iolist_t *payload,
+    bool ack = lorawan_hdr_get_ack(lw_hdr);
+    uint16_t conf_fcnt = 0;
+    (void)ack;
+    gnrc_lorawan_calculate_mic_uplink_11x(&mac->dev_addr, mac->mcps.fcnt, conf_fcnt,
```
ditto

jia200x commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_region.c:
```diff
@@ -106,6 +107,8 @@ uint32_t gnrc_lorawan_pick_channel(gnrc_lorawan_t *mac)
         state = bitarithm_test_and_clear(state, &index);
     }
+
+    mac->last_chan_idx = index;
```
since this function only picks a random channel, I would be in favor of maintaining "mac->last_chan_idx" outside of this function. It makes it easier to maintain as well

Ollrogge replied:
I am now managing it inside `gnrc_lorawan_mcps_request` by just looping through the channels and looking for the correct one. With this change I had removed the call to `gnrc_lorawan_pick_channel` from `gnrc_lorawan_event_retrans_timeout` which I had added because otherwise the `last_chan_idx` wont be correct.

I am wondering if this is bad since `gnrc_lorawan_event_retrans_timeout` might wan't to use a new channel ?

jia200x replied:
> I am wondering if this is bad since gnrc_lorawan_event_retrans_timeout might wan't to use a new channel ?

Hmmmm yes, the retransmssion should follow the same rules of any transmission. So it should choose a different channel.

Ollrogge replied:
Due to this I now moved the channel picking + the calculation of the uplink mic into the `_transmit_pkt` function. This is the cleanest solution I could find.

jia200x commented on sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:
```diff
@@ -28,20 +28,25 @@
-static uint8_t _nwkskey[LORAMAC_NWKSKEY_LEN];
 static uint8_t _appskey[LORAMAC_APPSKEY_LEN];
+#if IS_USED(CONFIG_LORAWAN_1_1)
 static uint8_t _appkey[LORAMAC_APPKEY_LEN];
```
why is `_appkey` not required in LoRaWAN 1.0.3?

Ollrogge replied:
It is but its name has changed. The 1.1 spec uses the NwkKey in place for the AppKey.

jia200x replied:
ok, I see. But maybe you could try to use the same variable names internally. Otherwise you cannot use regular `if (IS_ACTIVE())` stuff

jia200x commented on sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:
```diff
@@ -264,22 +273,54 @@ static int _init(gnrc_netif_t *netif)
     _set_be_addr(&netif->lorawan.mac, _devaddr);
-    gnrc_lorawan_init(&netif->lorawan.mac, netif->lorawan.nwkskey, netif->lorawan.appskey);
+
+#if IS_ACTIVE(CONFIG_LORAWAN_1_1)
+    const gnrc_lorawan_ctx_t ctx = {
+        .joineui = netif->lorawan.joineui,
+        .appskey = netif->lorawan.appskey,
+        .fnwksintkey = netif->lorawan.fnwksintkey,
+        .snwksintkey = netif->lorawan.snwksintkey,
+        .nwksenckey = netif->lorawan.nwksenckey,
+        .jsintkey = netif->lorawan.jsintkey,
+        .jsenckey = netif->lorawan.jsenckey
+    };
+
+    gnrc_lorawan_init(&netif->lorawan.mac, &ctx);
+#else
+    const gnrc_lorawan_ctx_t ctx = {
+        .joineui = netif->lorawan.joineui,
+        .appskey = netif->lorawan.appskey,
+        .fnwksintkey = netif->lorawan.fnwksintkey,
+    };
+
+    gnrc_lorawan_init(&netif->lorawan.mac, &ctx);
+#endif
```
```suggestion
    const gnrc_lorawan_ctx_t ctx = {
        .joineui = netif->lorawan.joineui,
        .appskey = netif->lorawan.appskey,
        .fnwksintkey = netif->lorawan.fnwksintkey,
 #if IS_ACTIVE(CONFIG_LORAWAN_1_1)
        .snwksintkey = netif->lorawan.snwksintkey,
        .nwksenckey = netif->lorawan.nwksenckey,
        .jsintkey = netif->lorawan.jsintkey,
        .jsenckey = netif->lorawan.jsenckey
 #endif
    };

    gnrc_lorawan_init(&netif->lorawan.mac, &ctx);
```

jia200x commented:
Overall the code looks nice and the changes make sense. I left some comments there and plan to run some tests when they are addressed.

jia200x commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -27,11 +27,25 @@
+ * If the end-device can be power-cycled then DevNonce SHALL be persistent
+ * (stored in a non-volatile memory)
+ */
+static uint16_t _dev_nonce = 0x0;
```
Is there a reason to put this here rather than in the MAC descriptor? (`gnrc_lorawan_t`).

Ollrogge replied:
No actually not. Removed it

Ollrogge commented:
> Overall the code looks nice and the changes make sense. I left some comments there and plan to run some tests when they are addressed.

I think I have addressed all comments now :)

jia200x commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mcps.c:
```diff
@@ -26,15 +26,15 @@
 #include "random.h"
 
-#define ENABLE_DEBUG      0
+#define ENABLE_DEBUG      1
```
Please disable DEBUG here

jia200x commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mcps.c:
```diff
@@ -44,8 +44,25 @@ static int gnrc_lorawan_mic_is_valid(uint8_t *buf, size_t len, uint8_t *nwkskey)
+        * modulo 2^16 of the “confirmed” uplink frame that is being acknowledged.
+        * In all other cases ConfFCnt = 0x0000.
+        */
+        bool ack = lorawan_hdr_get_ack(lw_hdr);
```
Could you implement that? Otherwise confirmed downlink message won't seem to work

jia200x commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mcps.c:
```diff
@@ -321,8 +338,30 @@ static void _transmit_pkt(gnrc_lorawan_t *mac)
         last_snip = last_snip->iol_next;
     }
 
+    uint32_t channel = gnrc_lorawan_pick_channel(mac);
+    (void)channel;
+
+    uint16_t conf_fcnt = 0;
+    if (IS_ACTIVE(CONFIG_GNRC_LORAWAN_1_1)) {
+        for (unsigned i = 0; i < GNRC_LORAWAN_MAX_CHANNELS; i++) {
+            if (mac->channel[i] == channel) {
+            mac->last_chan_idx = i;
+            }
```
since you are already there: maybe at this point it makes sense to simply change the `gnrc_lorawan_pick_channel` function to return the channel index. You can e.g add another helper function to get the channel frequency from the channel index. That way you don't need to run this loop again to find the channel index.

jia200x commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -27,11 +27,11 @@
 #include "net/lorawan/hdr.h"
 
-#define ENABLE_DEBUG 0
+#define ENABLE_DEBUG 1
```
Please disable debug

jia200x commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -40,31 +40,47 @@ static void _build_join_req_pkt(uint8_t *appeui, uint8_t *deveui,
+    * If the end-device can be power-cycled then DevNonce SHALL be persistent
+    * (stored in a non-volatile memory)
+    */
+    ((le_uint16_t *)mac->mlme.dev_nonce)->u16++;
```
since we are using `le_xxx` structures to make the code portable, we should probably also keep these operations portable. Therefore, please use `byteorder_xxx` functions here instead of just operating on the u16 member.

jia200x commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -111,25 +128,31 @@ void gnrc_lorawan_mlme_process_join(gnrc_lorawan_t *mac, uint8_t *data,
-    gnrc_lorawan_generate_session_keys(ja_hdr->app_nonce, mac->mlme.dev_nonce,
-                                       mac->appskey, mac->nwkskey,
-                                       mac->appskey);
+    if (IS_ACTIVE(CONFIG_GNRC_LORAWAN_1_1)) {
+        gnrc_lorawan_generate_session_keys(ja_hdr->join_nonce,
+                                        mac->mlme.dev_nonce, mac->joineui, mac);
+    }
+    else {
+        gnrc_lorawan_generate_session_keys(ja_hdr->join_nonce,
+                                        mac->mlme.dev_nonce, NULL, mac);
+    }
```
```suggestion
void *joineui = IS_ACTIVE(CONFIG_GNRC_LORAWAN_1_1) ? mac->joineui : NULL;
gnrc_lorawan_generate_session_keys(ja_hdr->join_nonce, mac->mlme.dev_nonce, joineui, mac);

```

jia200x commented on sys/net/gnrc/link_layer/lorawan/include/gnrc_lorawan_internal.h:
```diff
@@ -112,19 +115,32 @@ extern "C" {
+ * @brief MLME Join Request type
+ */
+typedef enum {
+    REJOIN_REQ_0,
+    REJOIN_REQ_1,
+    REJOIN_REQ_2,
+    JOIN_REQ = 0xFF,
```
Doxygen will probably complain here because of missing member documentation

jia200x commented on sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:
```diff
@@ -437,101 +474,130 @@ static int _set(gnrc_netif_t *netif, const gnrc_netapi_opt_t *opt)
+                         LORAMAC_NWKSKEY_LEN);
+        break;
+#else
+    case NETOPT_LORAWAN_NWKSKEY:
+    case NETOPT_LORAWAN_FNWKSINTKEY:
+    case NETOPT_LORAWAN_SNWKSINTKEY:
+    case NETOPT_LORAWAN_NWKSENCKEY:
```
General question: How does a user with LoRaWAN 1.1 firmware connects to a LoRaWAN 1.0.3 NS?

jia200x reviewed (changes requested):
Here's some round of more detailed comments.
As stated before, I think the code looks in general in good shape. I just have 2 general comments/questions:
1. Some variables seem to have a different meaning regarding the LoRaWAN version. This should be at least documented or preferably try to avoid diverging variables.
2. How does the user join to a LoRaWAN 1.0.x NS if using the 1.1 binary?

jia200x commented:
> That is not possible at the moment. I could however add e.g. a fall_back variable to the context struct and execute 1.0 code with 1.1 firmware if this flag is set.

Hmmm from what I read in the standard it only seems to be a matter of choosing the "right" keys in order to join an 1.0.x backend. All the rest should be retrocompatible by design. If that doesn't work, maybe there's something missing in the implementation.

Ollrogge commented:
@jia200x Should I rather prioritize stack usage or coverage ? E.g. should I leave a struct member only needed in LoRaWAN 1.1 to increase coverage with `if(IS_USED(...` or remove it with `#if` if using LoRaWAN 1.0 ?

jia200x commented:
Ideally members should be removed from the struct if they are not used.
Although this rules out if(IS_USED), you can always write getters/setters that map to no op if you don't need the member. You can then use if(IS_USED) with the getter/setter and use #ifdef in the getter/setter definition.

Ollrogge commented:
> Ideally members should be removed from the struct if they are not used. Although this rules out if(IS_USED), you can always write getters/setters that map to no op if you don't need the member. You can then use if(IS_USED) with the getter/setter and use #ifdef in the getter/setter definition.

In order for 1.1 firmware to work 1.0x NS I had to add handling of the `OptNeg` bit. This would have removed all possibilities to use `if(IS_USED)`. Therefore, I added a getter and setter for this. 

Furthermore I implemented the acknowledgement frame counter and addressed all other comments :)

jia200x commented on examples/gnrc_lorawan/Makefile:
```diff
@@ -13,7 +13,9 @@ USEMODULE += netdev_default
-USEMODULE += gnrc_lorawan
+#USEMODULE += gnrc_lorawan
+# Add support for GNRC LoRaWAN 1.1
+USEMODULE += gnrc_lorawan_1_1
```
Are `gnrc_lorawan` and `gnrc_lorawan_1_1` mutually exclusive?

Ollrogge replied:
No, `gnrc_lorawan_1_1` depends on `gnrc_lorawan`

jia200x commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mcps.c:
```diff
@@ -129,6 +138,23 @@ int gnrc_lorawan_parse_dl(gnrc_lorawan_t *mac, uint8_t *buf, size_t len,
+    uint32_t _fcnt =
+        gnrc_lorawan_fcnt_stol(mac->mcps.fcnt_down, _hdr->fcnt.u16);
+
+    uint32_t fcnt_down = mac->mcps.fcnt_down;
+#if IS_USED(MODULE_GNRC_LORAWAN_1_1)
+    if (pkt->port) {
+        fcnt_down = mac->mcps.afcnt_down;
+    }
+#endif
```
We should really avoid these #if macros here. If you can create a function `gnrc_lorawan_afcnt_down` and put there the #if macro it would simplify this snippet.

jia200x commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mcps.c:
```diff
@@ -175,7 +201,19 @@ void gnrc_lorawan_mcps_process_downlink(gnrc_lorawan_t *mac, uint8_t *psdu,
                                      key);
     }
 
+#if IS_USED(MODULE_GNRC_LORAWAN_1_1)
+    if (_pkt.port) {
+        mac->mcps.afcnt_down = _pkt.fcnt_down;
+    }
+    else {
+        mac->mcps.fcnt_down = _pkt.fcnt_down;
+    }
+
+    mac->mcps.last_fcnt_down = _pkt.fcnt_down;
+#else
     mac->mcps.fcnt_down = _pkt.fcnt_down;
+#endif
```
same here. Please avoid #if macros. In places where it's unavoidable (struct members), please use getters/setters with NO-OPs

jia200x commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -40,31 +40,49 @@ static void _build_join_req_pkt(uint8_t *appeui, uint8_t *deveui,
+                                          uint8_t *eui, uint8_t *key, uint8_t dr)
 {
+
+#if IS_USED(MODULE_GNRC_LORAWAN_1_1)
+    /**
+    * DevNonce starting at 0 when device is powered up and incremented with every
+    * Join-request.
+    *
+    * TODO: A DevNonce value SHALL NEVER be reused for a given JoinEUI value.
+    * If the end-device can be power-cycled then DevNonce SHALL be persistent
+    * (stored in a non-volatile memory)
+    */
+    uint16_t dev_nonce = byteorder_lebuftohs(mac->mlme.dev_nonce);
+    byteorder_htolebufs(mac->mlme.dev_nonce, dev_nonce++);
+
+    gnrc_lorawan_generate_lifetime_session_keys(deveui, mac->nwksenckey,
+                                               mac->jsintkey, mac->jsenckey);
+#else
```
please replace #if with if(IS_USED).

jia200x commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c:
```diff
@@ -237,43 +262,48 @@ void gnrc_lorawan_mlme_request(gnrc_lorawan_t *mac,
             break;
+        }
+
+#if IS_USED(MODULE_GNRC_LORAWAN_1_1)
+        memcpy(mac->appskey, mlme_request->join.appkey, LORAMAC_APPKEY_LEN);
+#endif
```
please use if(IS_USED) here. Although there's an overhead of adding a getter for join->appkey, we catch for free compilation errors.
Also, it doesn't seem to be the first part where the code requires these struct members. So, it's IMO definitely worth it to use getters here.

jia200x commented on sys/net/gnrc/link_layer/lorawan/include/gnrc_lorawan_internal.h:
```diff
@@ -147,7 +161,11 @@ typedef struct {
     uint32_t fcnt;                      /**< uplink framecounter */
-    uint32_t fcnt_down;                 /**< downlink frame counter */
+    uint32_t fcnt_down;                 /**< downlink frame counter. Mapped to network downlink frame counter if using LoRaWAN 1.1 */
+#if IS_USED(MODULE_GNRC_LORAWAN_1_1)
+    uint32_t afcnt_down;                /**< application downlink frame counter */
+    uint32_t last_fcnt_down;            /**< frame counter used to decrypt last frame. Needed to compute ConfFCnt */
+#endif
```
here and below: please add setters/getters for these members, in order to be able to use if(IS_USED) everywhere.

jia200x commented on sys/net/gnrc/link_layer/lorawan/include/gnrc_lorawan_internal.h:
```diff
@@ -488,6 +573,51 @@ void gnrc_lorawan_set_rx2_dr(gnrc_lorawan_t *mac, uint8_t rx2_dr);
+ * @return true if OptNeg bit is set
+ * @return false if OptNeg bit is unset
+ */
+static inline bool gnrc_lorawan_optneg_is_set(gnrc_lorawan_t *mac)
+{
+    (void) mac;
+#if IS_USED(MODULE_GNRC_LORAWAN_1_1)
+    return mac->optneg;
+#else
+    return 1;
+#endif
+}
```
This kind of setters/getters are the one I mean with the comment above :P

jia200x commented on sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:
```diff
@@ -28,20 +28,23 @@
 #include "net/gnrc/netreg.h"
 
-#define ENABLE_DEBUG 0
+#define ENABLE_DEBUG 1
```
Please disable debug

jia200x commented on sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:
```diff
@@ -264,22 +271,52 @@ static int _init(gnrc_netif_t *netif)
+    memcpy(netif->lorawan.fnwksintkey, _fnwksintkey, sizeof(_fnwksintkey));
+
+#if IS_USED(MODULE_GNRC_LORAWAN_1_1)
+    memcpy(netif->lorawan.snwksintkey, _snwksintkey, sizeof(_snwksintkey));
+    memcpy(netif->lorawan.nwksenckey, _nwksenckey, sizeof(_nwksenckey));
     memcpy(netif->lorawan.appkey, _appkey, sizeof(_appkey));
-    _memcpy_reversed(netif->lorawan.appeui, _appeui, sizeof(_appeui));
+#endif
```
please replace #if with if(IS_USED)

jia200x commented on sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:
```diff
@@ -437,101 +474,134 @@ static int _set(gnrc_netif_t *netif, const gnrc_netapi_opt_t *opt)
+                mlme_request.join.deveui = netif->lorawan.deveui;
+                mlme_request.join.joineui = netif->lorawan.joineui;
+                mlme_request.join.nwkkey = netif->lorawan.nwkkey;
+#if IS_USED(MODULE_GNRC_LORAWAN_1_1)
+                mlme_request.join.appkey = netif->lorawan.appkey;
+#endif
```
ditto

jia200x commented on tests/gnrc_lorawan_11/Makefile:
```diff
@@ -0,0 +1,13 @@
+# Add unittest framework
+USEMODULE += embunit
+
+USEMODULE += gnrc_lorawan
+
+CFLAGS += -DCONFIG_GNRC_LORAWAN_1_1
```
```suggestion
USEMODULE += gnrc_lorawan_1_1
```

jia200x commented on sys/net/gnrc/link_layer/lorawan/Kconfig:
```diff
@@ -16,6 +16,10 @@ menuconfig KCONFIG_USEMODULE_GNRC_LORAWAN
 
 if KCONFIG_USEMODULE_GNRC_LORAWAN
 
+config GNRC_LORAWAN_1_1
+    bool "Use LoraWAN version 1.1x"
+    default n
+
```
In an offline discussion with @leandrolanzieri we realized GNRC LoRaWAN is still not modeled in the dependency resolution. Therefore, we cannot include it as pseudomodule from Kconfig.

You have two options here
1. Either try to model the dependency resolution of Kconfig
2. Remove this symbol from Kconfig

jia200x commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan.c:
```diff
@@ -76,10 +77,21 @@ static void _sleep_radio(gnrc_lorawan_t *mac)
+    mac->joineui = ctx->joineui;
+    mac->appskey = ctx->appskey;
+
+#if IS_USED(MODULE_GNRC_LORAWAN_1_1)
+    mac->fnwksintkey = ctx->fnwksintkey;
+    mac->snwksintkey = ctx->snwksintkey;
+    mac->nwksenckey = ctx->nwksenckey;
+    mac->jsintkey = ctx->jsintkey;
+    mac->jsenckey = ctx->jsenckey;
+#else
+    mac->fnwksintkey = mac->snwksintkey = mac->nwksenckey = ctx->fnwksintkey;
+#endif
+
```
It seems all members of `gnrc_lorawan_ctx_t` end up in the MAC descriptor. Therefore, have you considered storing the pointer there? Then you can simply do:
```suggestion
    memcpy(&mac->ctx, ctx, sizeof(gnrc_lorawan_ctx_t));
```

That way you remove the #if macro here. Of course you will need to set all `ctx` members to `ctx->fnwksintkey`, but you can do that before calling `gnrc_lorawan_init`

jia200x reviewed (changes requested):
Last round of (style) comments.
Otherwise, I will proceed testing it against a Chirpstack instance.

From the Kconfig side I'm also not sure how to deal with the pseudomodule, since some of the options are exposed only if GNRC_LORAWAN_1_1 is used, but GNRC LoRaWAN is still not modeled. Maybe @leandrolanzieri could provide some insights?

leandrolanzieri commented on sys/net/link_layer/Kconfig.lorawan:
```diff
@@ -99,17 +99,37 @@ config LORAMAC_DEV_EUI_DEFAULT
+    help
+        Configure the default application key for join procedure.
+
+if GNRC_LORAWAN_1_1
```
As GNRC LoRaWAN is still not modelled in Kconfig, and we are moving the 1.1 to a pseudomodule, we can use here the `USEMODULE_`-prefixed symbol
```suggestion
if USEMODULE_GNRC_LORAWAN_1_1
```

Ollrogge commented on sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mcps.c:
```diff
@@ -44,8 +46,26 @@ static int gnrc_lorawan_mic_is_valid(uint8_t *buf, size_t len, uint8_t *nwkskey)
+        * modulo 2^16 of the “confirmed” uplink frame that is being acknowledged.
+        * In all other cases ConfFCnt = 0x0000.
+        */
+        if (lorawan_hdr_get_ack(lw_hdr)) {
+            conf_fnct = fcnt_up % 0x10000;
```
Why did you change this to a bitwise AND ? The spec specifically mentions: ` ConfFCnt is the frame counter value modulo 2^16 `

jia200x replied:
The bitwise AND is equivalent to the modulo operator in this context. But if `conf_fcnt` is a `uint16_t`, it should be fine to remove any operator

Ollrogge commented:
> Last round of (style) comments. Otherwise, I will proceed testing it against a Chirpstack instance.
> 
> From the Kconfig side I'm also not sure how to deal with the pseudomodule, since some of the options are exposed only if GNRC_LORAWAN_1_1 is used, but GNRC LoRaWAN is still not modeled. Maybe @leandrolanzieri could provide some insights?

Addressed all your latest comments now :)

Ollrogge commented:
@jia200x Do you need anything else to start testing?

jia200x commented:
> @jia200x Do you need anything else to start testing?

Time :smile: 

I will give it a look during the week

Ollrogge commented:
> > @jia200x Do you need anything else to start testing?
> 
> Time smile
> 
> I will give it a look during the week

All good :) Just wondered if there was maybe something I missed

jia200x commented:
> I will give it a look during the week

Uh, I couldn't make it this week :/. I will be back on Wednesday and plan to continue with testing

jia200x commented:
while testing, I realized the netif commands are not implemented in sc_gnrc_netif (so, you cannot e.g set the join key from the shell). This is important, since RIOTCtrl uses the shell for setting the keys.
Could you implement that as well? It should be straightforward

jia200x commented:
Otherwise I can join the Network Server using 1.1!
```
ifconfig
Iface  3  HWaddr: 26:0B:79:1B  Frequency: 868299987Hz  RSSI: -157  BW: 125kHz  SF: 12  CR: 4/5  Link: up 
           TX-Power: 14dBm  State: SLEEP  Demod margin.: 0  Num gateways.: 0 
          IQ_INVERT  
          RX_SINGLE  OTAA  L2-PDU:767  
```

jia200x commented:
after flashing again, I get these messages from the join server:
```
DevNonce is too small
```

Any ideas?

Ollrogge commented:
> while testing, I realized the netif commands are not implemented in sc_gnrc_netif (so, you cannot e.g set the join key from the shell). This is important, since RIOTCtrl uses the shell for setting the keys. Could you implement that as well? It should be straightforward

Done. Hope I didn't mess anything up. These key names are really hard to remember.

Ollrogge commented:
> after flashing again, I get these messages from the join server:
> 
> ```
> DevNonce is too small
> ```
> 
> Any ideas?

Yes this is due to the new requirements for the DevNonce in LoRaWAN 1.1: "A DevNonce value SHALL NEVER be reused for a given JoinEUI value. If the end-device can be power-cycled then DevNonce SHALL be persistent (stored in a non-volatile memory)".

This will require the LoRaWAN code to store state in flash memory. I will implement this tomorrow.

jia200x commented on sys/net/gnrc/link_layer/lorawan/include/gnrc_lorawan_internal.h:
```diff
@@ -99,6 +99,14 @@ extern "C" {
 #define GNRC_LORAWAN_FOPTS_MAX_SIZE (15U)               /**< Maximum size of Fopts field */
 #define GNRC_LORAWAN_FPORT_SIZE (1U)                    /**< Size of the Fport field */
 
+#if defined(CPU_FAM_STM32F2) || defined(CPU_FAM_STM32F4) || \
```
what's the rationale of this? Why do you treat these platforms differently?

Ollrogge replied:
These CPU's have different flashpage sizes. `FLASHPAGE_NUMOF` does not account for this and therefore using `FLASHPAGE_NUMOF -0x4` would not work since that flashpage doesn't exist.

benpicco replied:
How many flash pages do you need?

Ollrogge replied:
1

benpicco replied:
So why not always use the last one?

Ollrogge replied:
How ? Using `flashpage_last_free` ?

benpicco replied:
Why not use `FLASH_WRITABLE_INIT`?

Ollrogge replied:
Because it requires the `periph_flashpage_in_address_space` feature which would prevent boards like the `nucleo-f411re` from being able to use LoRaWAN 1.1

Ollrogge commented:
> hmmmm for some reason the device joins (TTN), but the NS does not receive data using the `txtsnd` command :(. I tried the same commands with LoRaWAN 1.0.3 (current master) and it works. Have you seen this behavior before? Are you using Chirpstack as NS?

No, unfortunately I haven't seen this before and yes I am using Chirpstack as NS. 
How are you testing the implementation with TTN ? Maybe I can reproduce and debug the issue.

jia200x commented:
> No, unfortunately I haven't seen this before and yes I am using Chirpstack as NS.
How are you testing the implementation with TTN ? Maybe I can reproduce and debug the issue.

I compile the `examples/gnrc_lorawan` with these settings (I X'd the keys)
```
diff --cc examples/gnrc_lorawan/Makefile
index d8cdb15b35,d8cdb15b35..5319cf4277
--- a/examples/gnrc_lorawan/Makefile
+++ b/examples/gnrc_lorawan/Makefile
@@@ -15,7 -15,7 +15,7 @@@ USEMODULE += auto_init_gnrc_neti
  # Add support for GNRC LoRaWAN
  USEMODULE += gnrc_lorawan
  # Add support for GNRC LoRaWAN 1.1
--# USEMODULE += gnrc_lorawan_1_1
++USEMODULE += gnrc_lorawan_1_1
  
  # Use GNRC pktdump to print downlink messages
  USEMODULE += gnrc_pktdump
@@@ -56,11 -56,11 +56,11 @@@ ifndef CONFIG_KCONFIG_USEMODULE_LORAWA
    # header. Therefore this flag will be removed after that
    CFLAGS += -DCONFIG_GNRC_NETIF_LORAWAN_NETIF_HDR
  
--  CFLAGS += -DCONFIG_LORAMAC_APP_KEY_DEFAULT=\"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\"
--  CFLAGS += -DCONFIG_LORAMAC_NWK_KEY_DEFAULT=\"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\"
--  CFLAGS += -DCONFIG_LORAMAC_APP_EUI_DEFAULT=\"BBBBBBBBBBBBBBBB\"
--  CFLAGS += -DCONFIG_LORAMAC_JOIN_EUI_DEFAULT=\"BBBBBBBBBBBBBBBB\"
--  CFLAGS += -DCONFIG_LORAMAC_DEV_EUI_DEFAULT=\"CCCCCCCCCCCCCCCC\"
++  CFLAGS += -DCONFIG_LORAMAC_APP_KEY_DEFAULT=\"XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\"
++  CFLAGS += -DCONFIG_LORAMAC_NWK_KEY_DEFAULT=\"XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
++  CFLAGS += -DCONFIG_LORAMAC_APP_EUI_DEFAULT=\"0000000000000000\"
++  CFLAGS += -DCONFIG_LORAMAC_JOIN_EUI_DEFAULT=\"0000000000000000\"
++  CFLAGS += -DCONFIG_LORAMAC_DEV_EUI_DEFAULT=\"XXXXXXXXXXXXXXXX\"
  
    # For TTN, It's necessary to set the RX2 DR to 3 in EU_868 region
    # CFLAGS += -DCONFIG_LORAMAC_DEFAULT_RX2_DR_3
```

The NS is configured with LoRaWAN 1.1 and "RP001 Regional Parameters 1.0.3 revision A". I simply run:
```
ifconfig 3 up
/* wait some seconds */
ifconfig /* <- shows link is up */
txtsnd 3 01 RIOT
```

The NS does not receive data.

jia200x commented:
> How do I access a device that is part of TTN ? I have never interacted with it

You need to register to TTN. You can do it here: https://eu1.cloud.thethings.network
Then, you need to add an Application and a Device (similar to Chirpstack). Since TTN works with "ready-to-use" devices, you will need to specify a devkit when selecting the device type. I usually select STM32 and any of the Nucleo-wl5x series.

Do you have a TTN gateway nearby? If so, you can just flash it to a RIOT device. Otherwise, you can use [IoT-LAB](https://www.iot-lab.info/) to book a `b-l072z-lrwan1` and work directly with the testbed. They have a TTN gateway there.

If you go with IoT-LAB, you can use the built-in iotlab Makefile logic on RIOT to interact with the remote nodes. Eg
```
IOTLAB_NODE=st-lrwan1-11.saclay.iot-lab.info BOARD=b-l072z-lrwan1 make clean flash term /* <--- voilá, the remote node is flashed and you get a shell */
```

Ollrogge commented:
> > How do I access a device that is part of TTN ? I have never interacted with it
> 
> You need to register to TTN. You can do it here: https://eu1.cloud.thethings.network Then, you need to add an Application and a Device (similar to Chirpstack). Since TTN works with "ready-to-use" devices, you will need to specify a devkit when selecting the device type. I usually select STM32 and any of the Nucleo-wl5x series.
> 
> Do you have a TTN gateway nearby? If so, you can just flash it to a RIOT device. Otherwise, you can use [IoT-LAB](https://www.iot-lab.info/) to book a `b-l072z-lrwan1` and work directly with the testbed. They have a TTN gateway there.
> 
> If you go with IoT-LAB, you can use the built-in iotlab Makefile logic on RIOT to interact with the remote nodes. Eg
> 
> ```
> IOTLAB_NODE=st-lrwan1-11.saclay.iot-lab.info BOARD=b-l072z-lrwan1 make clean flash term /* <--- voilà, the remote node is flashed and you get a shell */
> ```
> 
> EDIT: Please let me know if you need anything

So after some debugging I am 99% certain that no data is received because I haven't yet implemented the `RekeyInd` MAC command. I found [this](https://github.com/TheThingsNetwork/lorawan-stack/issues/2171#issuecomment-607164228) comment on an issue which indicates that TTN will drop uplink messages if `RekeyInd` hasn't been sent. Will implement this.

Ollrogge commented:
@jia200x I implement the Rekey MAC command and successfully tested it with TTN. Please test again if you can :)

![image](https://user-images.githubusercontent.com/20866465/171188128-fc545ab3-7974-4470-b22e-0fb1628dd811.png)

Ollrogge commented:
> Hi. It works now! However, according to [this](https://www.iotnet.eu/wp/lora-alliance-released-lorawan-1-1-specification/?print=print), there should be also backward compatibility between a v1.1 device and a 1.0.3 network server. When I configure that on TTN, it throws invalid MIC. Does it work for you on Chirpstack?

Should be fixed as well now !

jia200x commented:
I can confirm this still works as expected when compiling LoRaWAN 1.0.3. I will proceed with all tests for LoRaWAN 1.1

jia200x commented:
ok, everything seems to work now

jia200x reviewed (approved):
Tested ACK. Please squash commits to make them more semantic (e.g "add support for GNRC LoRaWAN 1.1", "Adapt test". etc...)

Ollrogge commented:
@jia200x Tests for some boards fail because they don't implement they flashpage API which I need for LoRaWAN 1.1.
How should I handle this ?

Ollrogge commented:
@jia200x Can you explain to me the doc and Murdock failures ? Neither seem to be related to my code ...

jia200x commented:
There seemed to be an unrelated error. I retriggered the CI. Let's see what happens

Ollrogge commented:
Looks like the same issue :(

jia200x commented:
> Looks like the same issue :(

I just realized that the Murdock problem was because you added NETOPTs without updating `tests/unittests/tests-netopt`. You have to add the new NETOPTs there to make the test pass.
Let's try to fix that one first and then see what went on with the documentation

Ollrogge commented:
> > Looks like the same issue :(
> 
> I just realized that the Murdock problem was because you added NETOPTs without updating `tests/unittests/tests-netopt`. You have to add the new NETOPTs there to make the test pass. Let's try to fix that one first and then see what went on with the documentation

Thanks for this ! I should have checked the error message more closely... All green now !

jia200x commented:
finally! congrats @Ollrogge :tada: :tada: :tada:

Ollrogge commented:
> finally! congrats @Ollrogge tada tada tada

Yaaayy :) Thank you for all the help !!
